C compiler error when unwrapping Optional in an array

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 63f5f4a, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 7af8faf.63f5f4a
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.2, 24C101
|Processor           |8 cpus, 64bit, little endian, Apple M2
|Memory              |0.06GB/16GB
|                    |
|V executable        |/Users/louis/Git/nv/v
|V last modified time|2025-01-10 15:29:47
|                    |
|V home dir          |OK, value: /Users/louis/Git/nv
|VMODULES            |OK, value: /Users/louis/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/louis/Git/v_staleworker
|                    |
|Git version         |git version 2.39.5 (Apple Git-154)
|V git status        |weekly.2025.1-29-g63f5f4a6-dirty (8 commit(s) behind V master)
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.4)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.4)
|clang version       |Homebrew clang version 18.1.8
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 713692d4
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  src/main_test.v && src/main_test`
```v
module main

fn make_option() ?string {
	return "abc"
}

fn test_rows() {
	cols := [make_option()]
	if col := cols[0] {
		println("I made it")
	}
}
```

**What did you see?**
```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/main_test.01JHB9CM93FXXWM4QG45C2FYHY.tmp.c:13055:6: error: use of undeclared identifier '_t1'
cc:  13055 |         if (_t1.state == 0) {
cc:        |             ^
cc: /tmp/v_501/main_test.01JHB9CM93FXXWM4QG45C2FYHY.tmp.c:13056:10: error: initializing 'string' (aka 'struct string') with an expression of incompatible type '_option_string' (aka 'struct _option_string')
cc:  13056 |                 string col = (*(_option_string*)array_get(cols, 0));
cc:        |                        ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc: 2 errors generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

I expected this code to compile. `cols[0]` is a `?string` and it should be unwrapped by `if val := col[0]`. I stumbled across this bug while using the `Row` type from `db.pg` library.


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

Ok, I pulled the latest commit but still get the same error:

```
% ~/Git/nv/v test src/main_test.v
---- Testing... ----------------------------------------------------------------------------------------------------------------------------
 FAIL     0.000 ms /Users/louis/Git/v_staleworker/src/main_test.v
>> compilation failed:
================== C compilation error (from cc): ==============
cc: /tmp/v_501/tsession_1ed3b0240_01JHBAFTBGQ8X3YGM1KJQY8JRH/main_test.01JHBAFV4X475V943FF2SXJGZW.tmp.c:12973:6: error: use of undeclared identifier '_t1'
cc:  12973 |         if (_t1.state == 0) {
cc:        |             ^
cc: /tmp/v_501/tsession_1ed3b0240_01JHBAFTBGQ8X3YGM1KJQY8JRH/main_test.01JHBAFV4X475V943FF2SXJGZW.tmp.c:12974:10: error: initializing 'string' (aka 'struct string') with an expression of incompatible type '_option_string' (aka 'struct _option_string')
cc:  12974 |                 string col = (*(_option_string*)array_get(cols, 0));
cc:        |                        ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc: 2 errors generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

V doctor:

```
% ~/Git/nv/v doctor              
|V full version      |V 0.4.9 7af8faf.81b421b
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.2, 24C101
|Processor           |8 cpus, 64bit, little endian, Apple M2
|Memory              |0.14GB/16GB
|                    |
|V executable        |/Users/louis/Git/nv/v
|V last modified time|2025-01-11 18:35:28
|                    |
|V home dir          |OK, value: /Users/louis/Git/nv
|VMODULES            |OK, value: /Users/louis/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/louis/Git/v_staleworker
|                    |
|Git version         |git version 2.39.5 (Apple Git-154)
|V git status        |weekly.2025.1-35-g81b421bb-dirty
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.4)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.4)
|clang version       |Homebrew clang version 18.1.8
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 713692d4
|emcc version        |N/A
|glibc version       |N/A
```

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -260,6 +260,15 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 					cond.expr_type
 				}
 				g.writeln('${g.styp(g.unwrap_generic(cond_expr_type))} ${var_name};')
+			} else if cond.expr is ast.IndexExpr {
+				value_type := g.table.value_type(g.unwrap_generic(cond.expr.left_type))
+				if value_type.has_flag(.option) {
+					var_name := g.new_tmp_var()
+					guard_vars[i] = var_name
+					g.writeln('${g.styp(value_type)} ${var_name};')
+				} else {
+					guard_vars[i] = ''
+				}
 			} else {
 				guard_vars[i] = ''
 			}

--- a/vlib/v/tests/options/option_ifguard_array_of_option_test.v
+++ b/vlib/v/tests/options/option_ifguard_array_of_option_test.v
@@ -4,7 +4,7 @@ fn make_option() ?string {
 	return 'abc'
 }
 
-fn main() {
+fn test_main() {
 	cols := [make_option()]
 	if col := cols[0] {
 		assert col == 'abc'

--- a/vlib/v/tests/options/option_ifguard_array_of_option_test.v
+++ b/vlib/v/tests/options/option_ifguard_array_of_option_test.v
@@ -1,0 +1,14 @@
+module main
+
+fn make_option() ?string {
+	return 'abc'
+}
+
+fn main() {
+	cols := [make_option()]
+	if col := cols[0] {
+		assert col == 'abc'
+		assert true
+		dump(col)
+	}
+}

